### PR TITLE
🐛 Add validation testing for defaulting

### PR DESCRIPTION
**Pull request by randomvariable**
**What this PR does / why we need it**:
Intended to prevent regressions in webhooks when additive changes are made to the API, add generic tests to all of the controllers with defaulting webhooks to ensure that they pass validations.

As a result, found an edge case in CRS on update which is the second commit here (rather than squashing into a single commit).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by randomvariable** [comment]
This PR will fail tests until rebased on #4439

**Comment by fabriziopandini** [comment]
@randomvariable changes sgtm to me.
Only two questions:
- should we generalise this into util like we are doing of the fuzzer tests?
- does this test are a duplicate of the test being introduced by https://github.com/kubernetes-sigs/cluster-api/pull/4439? should we dedup?

**Comment by randomvariable** [comment]
> should we generalise this into util like we are doing of the fuzzer tests?
I'll see if this is doable.

> does this test are a duplicate of the test being introduced by #4439? should we dedup?

Yes, we should dedup, but could do it in post.

**Comment by randomvariable** [comment]
Created a `defaulting` package until util that includes the new test helper.
Also, I had missed out CRS and MachinePool, so have added to those as well.

File: util/defaulting/defaulting.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// and validating webhooks.
+type DefaultingValidator interface {
+	admission.Defaulter
+	admission.Validator
+}
+
+// ValidateDefaultingTest returns a new testing function to be used in tests to
+// make sure defaulting webhooks also pass validation tests both on create
+// and update.
+func ValidateDefaultingTest(object DefaultingValidator) func(*testing.T) {
```
Comment (MarcelMue) [inline util/defaulting/defaulting.go:36]: Does it make sense to include `validate-on-delete` here? It is part of the validator interface - currently it's `NOOP` basically everywhere, but including it for completeness would make sense to me.

Comment (randomvariable) [reply util/defaulting/defaulting.go:36]: Good call. Added

File: api/v1alpha4/cluster_webhook_test.go [hunk 2]
```diff
@@ -43,6 +44,20 @@ func TestClusterDefault(t *testing.T) {
 	g.Expect(c.Spec.ControlPlaneRef.Namespace).To(Equal(c.Namespace))
 }
 
+func TestClusterDefaultingValidation(t *testing.T) {
+	c := &Cluster{
```
Comment (MarcelMue) [inline api/v1alpha4/cluster_webhook_test.go:48]: I don't love how we currently essentially have the same object in several test cases (see `TestClusterDefault` in this case). It would be nice to make this a bit more DRY - maybe something for a follow up.

File: util/defaulting/defaulting.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// DefaultingValidator interface is for objects that define both defaulting
+// and validating webhooks.
+type DefaultingValidator interface {
+	admission.Defaulter
+	admission.Validator
+}
+
+// ValidateDefaultingTest returns a new testing function to be used in tests to
+// make sure defaulting webhooks also pass validation tests both on create
+// and update.
```
Comment (MarcelMue) [inline util/defaulting/defaulting.go:35]:
```suggestion
// update and delete.
```

**Review by MarcelMue — COMMENTED** [review]
/LGTM

File: api/v1alpha4/cluster_webhook_test.go (hunk 2 above)
Comment (randomvariable) [reply api/v1alpha4/cluster_webhook_test.go]: Deduped it in place.

**Comment by randomvariable** [comment]
Found an edge case in CRS.

**Comment by randomvariable** [comment]
Looks like we're good now.

**Comment by jan-est** [comment]
looks good to me

**Comment by MarcelMue** [comment]
/LGTM
(for the third time I guess :D )

**Comment by randomvariable** [comment]
> (for the third time I guess :D )
I'm done now, I promise.

File: util/defaulting/defaulting.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+// and validating webhooks.
+type DefaultingValidator interface {
+	admission.Defaulter
+	admission.Validator
+}
+
+// ValidateDefaultingTest returns a new testing function to be used in tests to
+// make sure defaulting webhooks also pass validation tests on create,
+// update and delete.
+func ValidateDefaultingTest(object DefaultingValidator) func(*testing.T) {
```
Comment (fabriziopandini) [inline util/defaulting/defaulting.go:36]: Super nit: should this be `DefaultValidateTest` (it is more consistent with what the test is actually do and with the `DefaultingValidator` input type)

Comment (randomvariable) [reply util/defaulting/defaulting.go]: have renamed.

**Comment by fabriziopandini** [comment]
@randomvariable this PR introduced a useful safeguard; are you planning to backport this to the v0.3 branch as well?

**Comment by randomvariable** [comment]
> are you planning to backport this to the v0.3 branch as well?

I was about to ask if I should...
